kramer/src/main/kotlin/kramer/MavenRepo.kt

```diff
@@ -0,0 +1,331 @@
+
+  private val threadCount: Int by option("--threads").int().default(1)
+
+  internal val kontext by requireObject<Kontext>()
+
+  internal data class IndexEntry(
+    val dependants: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap()),
+    val versions: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap())
+  )
+
+  @FlowPreview
+  @ExperimentalCoroutinesApi
+  override fun run() {
+    val repoConfig = parseRepoConfig(configFile)
+    val count = AtomicInteger(0)
+    var exit = 0
+    val seen = ConcurrentHashMap<String, IndexEntry>()
+    val unresolved = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
+    val declaredArtifactSlugs = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
+    val issue = ConcurrentHashMap<String, String>()
+    val benchmark = measureTimeMillis {
+      if (repoConfig.artifacts.isNotEmpty()) Files.createDirectories(workspace)
+      runBlocking {
+        kontext.out(false) { "Building workspace for ${repoConfig.artifacts.size} artifacts" }
+        kontext.out { if (kontext.verbosity > 1) "using $threadCount threads." else "." }
+        repoConfig.artifacts.entries.asFlow()
+          .onEach {
+            // TODO: Validate per-artifact configuration.
+          }
+          .flatMapMerge(concurrency = threadCount) { (artifact_spec, config) ->
+            flow {
+              val resolver = ArtifactResolver(
+                cacheDir = kontext.localRepository,
+                suppressAddRepositoryWarnings = true,
+                repositories =
+                if (kontext.repositories.isNotEmpty()) kontext.repositories
+                else Repositories.DEFAULT
+              )
+              val artifact = resolver.artifactFor(artifact_spec)
+              val slug = "${artifact.groupId}:${artifact.artifactId}"
+              val entry = seen.getOrPut(slug) { IndexEntry() }
+              entry.versions.add(artifact.version)
+              declaredArtifactSlugs.add(slug)
+              var resolved: ResolvedArtifact? = null
+              val time = measureTimeMillis {
+                resolved = resolver.resolveArtifact(artifact)
+              }
+              val success = if (resolved != null) "successfully" else "unsuccessfully"
+              kontext.info { "Resolved ${artifact.coordinate} $success in ${time / 1000.0} seconds" }
+              resolved?.model?.apply {
+                kontext.verbose {
+                  dependencies.filter { it.scope != "test" }
+                    .joinToString("\n") { dep ->
+                      "  - ${dep.groupId}:${dep.artifactId}:${dep.version} - ${dep.scope}" +
+                        if (dep.exclusions.isNotEmpty()) {
+                          "\n    exclusions = ${dep.exclusions.forEach { ex -> ex.toString() }}"
+                        } else ""
+                    }
+                }
+              }
+              emit(artifact to resolved tre resolver fo config)
```

Comment:
could you define a type for this? It's not particularly readable as it stands.  It appears to be emitting the resolver with the artifact, resolved, and config. But for what purpose? What stage of resolution needs all these?

Reply:
resolver contains context which I want to follow the resolved for when I do artifact downloading to extract the .aar. As to a type, it's a quad-tuple. But I can make it cleaner. I'll use a real data class.

Reply:
Hmm. I actually want to push back. There is no type here, this is a tuple of four types, and the name/meaning associated iwth it would be... "ResolvedArtifactContext" or some crap like that, because it's really just a tuple. All the meaning you're going to get is in the emit statement, where I join together four things. It's only consumed in the next two flow steps (once as a filter, so only partially) and then it goes away. It doesn't really feel like it deserves a type. 

Instead of a real type, I'll move the recording of unresolved up out of the filter step, so I don't need artifact anymore. That should make it a bit less crappy.

Reply:
It's a step in the process, and each step can be expressed as a type. In fact, it's an excellent way to improve testability.

But I'm still not sure why resolver and config need to be returned?

Reply:
Config is the configuration for that artifact, which has informatino used in (for example) for dependency excludes. So I need it in the template step. 

I'm really sure I don't agree that a separate type is needed or adds testability since you'll only ever see its constructed, and it's destructured literally two or three lines below it's construction. It adds no semantic information over a tuple as far as I can tell.

At this point, it's not Quad, but Triple, which already exists in the kotlin sdk. 

I guess i"m not deeply opposed in theory - and the name "ResolutionResult" isn't bad. Where I end up stuck is that it adds a whole extra definition you end up going to look at, out of band, which doesn't add any more than a name, and maybe some type information if you aren't looking in an ide which expresses the type information in hints. But then you're having to look away from the code just to get ... what... nearly no extra context? It just doesn't feel like it carries its weight as a type, especially not in a flow like this.

Reply:
Oh, and as I said, resolver may keep state for caching (the way the model resolution system in maven is designed, which constrains maven-archeologist) and I wanted to reuse the resolver from the pom to download teh artifact. And because I need to strip out package from manifest from aar, I need to fetch the artifact. So I keep the resolver for an extra step. 

I could just create a new one, I suppose. But it's not really designed for that. I only have one per-flow-lane (so to speak) because i dont' want to share them between flatMapMerge calls because they're not themselves threadsafe.

Reply:
I consume the given `resolver` reference a few lines below  (125 in the current code).  Making it twice just seems like a bourgois excess. :)

Reply:
I was wrong, both about the intermediary types and extracting flow steps. It's not how I've seen this done much, but I think I have a reasonable approach to extracting the pipeline steps into separable, legible functions which can be wired together in the main flow, and I do end up using better-named types since now there is a formal signature to be reckoned with.